library/src-bootstrapped/scala/runtime/quoted/Matcher.scala

```diff
@@ -0,0 +1,276 @@
+//            else foldMatchings(paramss1.zip(paramss2).map { (params1, params2) => treesMatch(params1, params2) }: _*)
+//          val (bindMatch, tptMatch) = bindingAndTptMatch(scrutinee.symbol, tpt1, tpt2)
+//          val rhsEnv =
+//            env + (scrutinee.symbol -> pattern.symbol) ++
+//            typeParams1.zip(typeParams2).map((tparam1, tparam2) => tparam1.symbol -> tparam2.symbol) ++
+//            paramss1.flatten.zip(paramss2.flatten).map((param1, param2) => param1.symbol -> param2.symbol)
+//          val rhsMatch = treeMatches(rhs1, rhs2)(rhsEnv)
+//
+//          foldMatchings(bindMatch, typeParmasMatch, paramssMatch, tptMatch, rhsMatch)
+
+        case (Lambda(_, tpt1), Lambda(_, tpt2)) =>
+          // TODO match tpt1 with tpt2?
+          Some(())
+
+        case (Match(scru1, cases1), Match(scru2, cases2)) =>
+          val scrutineeMacth = treeMatches(scru1, scru2)
+          val casesMatch =
+            if (cases1.size != cases2.size) None
+            else foldMatchings(cases1.zip(cases2).map(caseMatches): _*)
+          foldMatchings(scrutineeMacth, casesMatch)
+
+        case (Try(body1, cases1, finalizer1), Try(body2, cases2, finalizer2)) =>
+          val bodyMacth = treeMatches(body1, body2)
+          val casesMatch =
+            if (cases1.size != cases2.size) None
+              else foldMatchings(cases1.zip(cases2).map(caseMatches): _*)
+          val finalizerMatch = treeOptMatches(finalizer1, finalizer2)
+          foldMatchings(bodyMacth, casesMatch, finalizerMatch)
+
+        case _ =>
+          if (debug)
+            println(
+              s""">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
+                 |Scrutinee
+                 |  ${scrutinee.showCode}
+                 |
+                 |${scrutinee.show}
+                 |
+                 |did not match pattern
+                 |  ${pattern.showCode}
+                 |
+                 |${pattern.show}
+                 |
+                 |
+                 |
+                 |
+                 |""".stripMargin)
+          None
+      }
+    }
+
+    def treeOptMatches(scrutinee: Option[Tree], pattern: Option[Tree])(implicit env: Set[(Symbol, Symbol)]): Option[Tuple] = {
+      (scrutinee, pattern) match {
+        case (Some(x), Some(y)) => treeMatches(x, y)
+        case (None, None) => Some(())
+        case _ => None
+      }
+    }
+
+    def caseMatches(scrutinee: CaseDef, pattern: CaseDef)(implicit env: Set[(Symbol, Symbol)]): Option[Tuple] = {
+      val (caseEnv, patternMatch) = patternMatches(scrutinee.pattern, pattern.pattern)
+      val guardMatch = treeOptMatches(scrutinee.guard, pattern.guard)(caseEnv)
+      val rhsMatch = treeMatches(scrutinee.rhs, pattern.rhs)(caseEnv)
+      foldMatchings(patternMatch, guardMatch, rhsMatch)
+    }
+
+    def patternMatches(scrutinee: Pattern, pattern: Pattern)(implicit env: Set[(Symbol, Symbol)]): (Set[(Symbol, Symbol)], Option[Tuple]) = (scrutinee, pattern) match {
```

Comment:
What gets returned here? Would be good to document that as well.

Reply:
I added the documentation